### PR TITLE
Calibrate regularizer

### PR DESCRIPTION
**Pull request by AlexLipp**
Adds a function `calibrate_regularizer` which sweeps regularizer parameter space and visualises trade-off plot allowing an optimal value to be chosen.

### CODE REVIEW

File: geochem_inverse_optimize.py [hunk 1]
```diff
@@ -245,6 +251,29 @@ def get_unique_upstream_areas(sample_network: nx.DiGraph):
...
+def calibrate_regularizer(
+    sample_network: nx.DiGraph,
+    element_data: ElementData,
+    min_: float,
+    max_: float,
+    number: float,
+):
+    vals = np.logspace(min_, max_, num=number)  # regularizer strengths to try
+    for val in vals:
+        print("________________________________________________________")
```
Comment (r-barnes) [inline geochem_inverse_optimize.py:263]:
```suggestion
        print(20*"_")
```

File: geochem_inverse_optimize.py [hunk 2]
```diff
@@ -245,6 +251,29 @@ def get_unique_upstream_areas(sample_network: nx.DiGraph):
...
+        print("Trying regularizer strength: 10^", round(np.log10(val), 3))
+        _, _ = sample_network.solve(element_data, solver="ecos", regularization_strength=val)
+        roughness = cp.norm(cp.vstack(sample_network._regularizer_terms)).value
```
Comment (r-barnes) [inline geochem_inverse_optimize.py:266]: Using underscored properties outside of their owner is an antipattern since that's an indicator that the thing is private. Consider adding a method to the class that gives you `cp.norm(cp.vstack(sample_network._primary_terms)).value` or what not and using that method here.

This separates the internal implementation of the class from its interface.

File: unmix_mwe.py [hunk 3]
```diff
@@ -15,7 +15,7 @@
 obs_data = pd.read_csv("data/geochem_no_dupes.dat", delimiter=" ")
 obs_data = obs_data.drop(columns=["Bi", "S"])
 
-element = "Mg"  # Set element
+element = "Be"  # Set element
```
Comment (r-barnes) [inline unmix_mwe.py:18]: Why this change in this PR?

File: unmix_mwe.py [hunk 4]
```diff
@@ -27,8 +27,13 @@
...
     element, obs_data
 )  # Return dictionary of {sample_name:concentration}
+
+
+gio.calibrate_regularizer(problem, element_data, -5, -1, 11)
+
+
 element_pred_down, element_pred_upstream = problem.solve(
-    element_data, solver="ecos", regularization_strength=10 ** (-3)
+    element_data, solver="ecos", regularization_strength=10**-3
```
Comment (r-barnes) [inline unmix_mwe.py:36]: Revert this inconsequential change?

File: geochem_inverse_optimize.py [hunk 5]
```diff
@@ -245,6 +251,29 @@ def get_unique_upstream_areas(sample_network: nx.DiGraph):
     return areas
 
 
+def calibrate_regularizer(
```
Comment (r-barnes) [inline geochem_inverse_optimize.py:254]: This function doesn't actually calibrate anything, it just makes a plot. Maybe change the name to reflect that, eg `plot_sweep_of_regularizer_strength`.

**Review by r-barnes — CHANGES_REQUESTED** [review]

File: geochem_inverse_optimize.py (hunk 2 above)
Comment (AlexLipp) [reply geochem_inverse_optimize.py:266]: Thanks for introducing to the concept of an antipattern, and this specific instance. 

I've added `get_roughness` and `get_misfit` methods to the class.

File: unmix_mwe.py (hunk 3 above)
Comment (AlexLipp) [reply unmix_mwe.py:18]: Removed.

File: geochem_inverse_optimize.py (hunk 5 above)
Comment (AlexLipp) [reply geochem_inverse_optimize.py:254]: Agreed. Changed as suggested.

File: geochem_inverse_optimize.py [hunk 6]
```diff
@@ -202,6 +208,12 @@ def solve(
 
         return downstream_preds, upstream_preds
 
+    def get_misfit(self) -> float:
+        return cp.norm(cp.vstack(self._primary_terms)).value
+
+    def get_roughness(self) -> float:
```
Comment (r-barnes) [inline geochem_inverse_optimize.py:214]: I'm not sure if `roughness` is the best name for what this is returning (or `misfit` for the above). I'll approve anyway.

**Review by r-barnes — APPROVED** [review]

**Comment by r-barnes** [comment]
@AlexLipp - I'll handle the merge, if that's okay.
